src/commands/build-util.ts

```diff
@@ -1,19 +1,126 @@
 import type CSSModuleLoader from 'css-modules-loader-core';
-import type {EventEmitter} from 'events';
-import execa from 'execa';
-import npmRunPath from 'npm-run-path';
-import {
-  BuildScript,
-  SnowpackConfig,
-  SnowpackPluginBuildArgs,
-  SnowpackPluginBuildResult,
-} from '../config';
+import {EventEmitter} from 'events';
+import path from 'path';
+import {SnowpackBuildMap, SnowpackConfig, SnowpackPlugin, SnowpackSourceFile} from '../config';
+import {getExt, replaceExt} from '../util';
+
+export interface BuildFileOptions {
+  buildPipeline: SnowpackPlugin[];
+  messageBus: EventEmitter;
+  isDev: boolean;
+}
 
 const IS_PREACT = /from\s+['"]preact['"]/;
 export function checkIsPreact(filePath: string, contents: string) {
   return filePath.endsWith('.jsx') && IS_PREACT.test(contents);
 }
 
+export function getInputsFromOutput(fileLoc: string, plugins: SnowpackPlugin[]) {
+  const ext = path.extname(fileLoc);
+  const potentialInputs = new Set([fileLoc]);
+  for (const plugin of plugins) {
+    if (plugin.output.includes(ext)) {
+      plugin.input.forEach((inp) => potentialInputs.add(fileLoc.replace(ext, inp)));
+    }
+  }
+  return Array.from(potentialInputs);
+}
+
+/** Core Snowpack file pipeline builder */
+export async function buildFile(
+  srcPath: string,
+  srcContents: string,
+  {buildPipeline, messageBus, isDev}: BuildFileOptions,
+): Promise<SnowpackBuildMap> {
+  const srcExt = getExt(srcPath);
+  const output: SnowpackBuildMap = {}; // important: clear output for each src file to keep memory low
+  output[srcPath] = {...srcExt, contents: srcContents, locOnDisk: srcPath}; // this is the object we’ll mutate through transformations
+
+  // Run the file through the Snowpack build pipeline:
+  for (const step of buildPipeline) {
+    // For now, we only run through build plugins that match at least one file extension. 
+    // All other plugins are ignored.
+    if (!step.build || step.input.length === 0) {
+      continue;
+    }
+    const staleOutputs = new Set<string>();
+    for (const destPath of Object.keys(output)) {
```

Comment:
Should we use `Object.entries(output)` here since we're using `destPath` to lookup the value later anyways?

`for (const [destPath, destBuildFile] of Object.entries(output)) {`